File: sort/heap_sort/Rust/heapsort.rs
// rust heapsort

pub fn heapsort(arr: &mut [i32]) {
	//heapify part
	//this procedure would build a valid max-heap (or min-heap for sorting descendantly)
	let end = arr.len();
	for start in (0..end / 2).rev() {
	    // Skip leaf nodes (end / 2).
	    sift_down(arr, start, end - 1);
	}

    //sorting part
	//iteratively sift down unsorted part (the heap).
	for end in (1..arr.len()).rev() {
	    arr.swap(end, 0);
	    sift_down(arr, 0, end - 1);
	}
}

//internal function for heap to fix itself to conform to heap definition.

//recondition: all elements below `start` are in heap order

//expect `start` itself.

fn sift_down(arr: &mut [i32], start: usize, end: usize) {
	let mut root = start;
	loop {
	    let mut child = root * 2 + 1; // Get the left child
	    if child > end {
	        break;
	    }
	    if child < end && arr[child] < arr[child + 1] {
	        //right child exists and is greater.
	        child += 1;
	    }

        if arr[root] < arr[child] {
	        //if child is greater than root, swap'em!
	        arr.swap(root, child);
	        root = child;
	    } else {
	        break;
	    }
	}
}

#[cfg(test)]

mod base {
	use super::*;
	base_cases!(heapsort);
}

File: sort/heap_sort/scala/HeapSort.scala
object HeapSort {

  def swap(arr: Array[Int], a: Int, b: Int) = {
    val aux = arr(a)
    arr(a) = arr(b)
    arr(b) = aux
  }

  def heapify(arr: Array[Int], n: Int, i: Int): Unit = {
    var largest = i
    val left = 2 * i + 1
    val right = 2 * i + 2

    if (left < n && arr(left) > arr(largest)) largest = left
    if (right < n && arr(right) > arr(largest)) largest = right

    if (largest != i) {
      swap(arr, i, largest)
      heapify(arr, n, largest)
    }
  }

  def heapSort(arr: Array[Int]) = {
    val sortedArr = arr.clone()
    val n = sortedArr.length

    for (i <- n / 2 - 1 to 0 by -1) {
      heapify(sortedArr, n, i)
    }

    for (i <- n - 1 to 0 by -1) {
      swap(sortedArr, 0, i)
      heapify(sortedArr, i, 0)
    }

    sortedArr
  }


  def main(args: Array[String]): Unit = {
    val arr = Array(3, 1, 4, 5, 9, 2, 6, 8, 7, 10)
    val sortedArray = heapSort(arr)
    println(sortedArray.mkString(", "))
  }
}
